### PR TITLE
wifi-subsys: Add wifi component

### PR DESCRIPTION
**Pull request by xkevin190**
### Contribution description

 Was added the wifi module this module was reused of [turpial firmware repo ](https://github.com/btcven/turpial-firmware/tree/v0.1.0-idf/components/Network) I proved the sta mode, but for this implementation sta this need more work because not working correctly, i did build the example of idf for sta and that example work. we could take this estructure but it need some fixes.

Other think is this implemntation uses many functions for only rename functions of idf  for example:

```c   
esp_err_t WiFi::setStaConfig(wifi_config_t& sta_config)
{
    return esp_wifi_set_config(WIFI_IF_STA, &sta_config);
}
```

I think that is not necessary  also  we could reduce much code, also do more in less steps. please review the code and share me your opinios.

this PR  still is in progress.

### CODE REVIEW

**Comment by xkevin190** [comment]
AP and STA mode was added correctly still is necessary add the functions for set configuration in the storage, meanwhile add some settings in the kconfig for start  and change ssid and password since kconfig, also turn on or turn off the sta or ap.

please check the code and share with me some doubt.

**Review by luisan00 — CHANGES_REQUESTED** [review]
static tests errors/warning must be fixed before

File: wifi-subsys/main/main.c [hunk 1]
```diff
@@ -22,19 +22,45 @@
  *
  */
 #include <stdio.h>
+#include <string.h>
 #include "sdkconfig.h"
 #include "freertos/FreeRTOS.h"
 #include "freertos/task.h"
 #include "esp_system.h"
+#include <esp_log.h>
```
Comment (luisan00) [inline wifi-subsys/main/main.c:30]: it's not a std library
```suggestion
#include "esp_log.h"
```

File: wifi-subsys/components/network/src/WiFi.c [hunk 2]
```diff
@@ -0,0 +1,185 @@
+
+#include "wifi.h"
+#include "esp_log.h"
+#include <esp_netif.h>
+#include <esp_err.h>
+#include <esp_event.h>
```
Comment (luisan00) [inline wifi-subsys/components/network/src/WiFi.c:6]: it's not part od std c libraries

**Comment by xkevin190** [comment]
there is some functions that uses nvs, although the nvs is not completed yet I  wanted to do the necessary functions.

File: wifi-subsys/components/network/src/WiFi.h [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+ * @return 1 when the wifi module is configured
+ *         0 when the wifi is not configured
+ */
+uint8_t wifi_credentials_are_configured();
+
+#ifdef __cplusplus
+}
+#endif
+#endif /* WIFI_H */
+/** @} */
```
Comment (luisan00) [inline wifi-subsys/components/network/src/WiFi.h:96]: no new line at EOF

File: wifi-subsys/components/network/test/TestWiFi.c [hunk 4]
```diff
@@ -0,0 +1,23 @@
+/**
+ * @file test_wifi_mode.cpp
+ * @author Locha Mesh Developers (contact@locha.io)
+ * @brief
+ * @version 0.1
+ * @date 2019-09-17
+ *
+ * @copyright Copyright (c) 2019 Locha Mesh project developers
+ * @license Apache 2.0, see LICENSE file for details
+ */
```
Comment (luisan00) [inline wifi-subsys/components/network/test/TestWiFi.c:10]: wrong header format

File: wifi-subsys/main/default_params.h [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+#define AP_WIFI_SSID "m4a_AP"           /*!< default AP ssid */
+#define AP_WIFI_PASSWORD "m4a-default"    /*!< default AP password */
+
+
+#ifdef __cplusplus
+}
+#endif
+#endif /* DEFAULT_PARAMS_H */
+/** @} */
```
Comment (luisan00) [inline wifi-subsys/main/default_params.h:62]: sqme here

File: wifi-subsys/main/main.c (hunk 1 above)
Comment (luisan00) [reply wifi-subsys/main/main.c:30]: this still untouched

**Review by luisan00 — CHANGES_REQUESTED** [review]
well done, exception of some little thing but good job in general

File: wifi-subsys/components/network/src/WiFi.c [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+#include "esp_err.h"
+#include "esp_event.h"
+#include "esp_system.h"
+#include "wifi.h"
+#include "storage.h"
+#include "default_params.h"
+
+static const char* TAG = "Wifi Module";
+
+
```
Comment (luisan00) [inline wifi-subsys/components/network/src/WiFi.c:37]: extra line xD

File: wifi-subsys/components/network/src/WiFi.c [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+            ESP_LOGI(TAG, "retry to connect to the AP");
+        }
+        ESP_LOGI(TAG,"connect to the AP fail");
+    } else if (event_base == IP_EVENT && event_id == IP_EVENT_STA_GOT_IP) {
+        ip_event_got_ip_t* event = (ip_event_got_ip_t*) event_data;
+        ESP_LOGI(TAG, "got ip:" IPSTR, IP2STR(&event->ip_info.ip));
+        s_retry_num = 0;
+    }
+
+
```
Comment (luisan00) [inline wifi-subsys/components/network/src/WiFi.c:59]: here too

File: wifi-subsys/components/network/src/WiFi.c [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+        return err;
+    }
+
+    ESP_LOGI(TAG, "wifi_init_softap finished. SSID:%s password:%s channel:%d",
+             wifi_config.ap.ssid, wifi_config.ap.password, AP_CHANNEL);
+
+    return ESP_OK;
+}
+
+
```
Comment (luisan00) [inline wifi-subsys/components/network/src/WiFi.c:153]: and more

File: wifi-subsys/components/network/src/WiFi.c [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+                                                        NULL,
+                                                        &instance_any_id));
+
+    ESP_ERROR_CHECK(esp_event_handler_instance_register(IP_EVENT,
+                                                        IP_EVENT_STA_GOT_IP,
+                                                        &event_handler,
+                                                        NULL,
+                                                        &instance_got_ip));
+
+
```
Comment (luisan00) [inline wifi-subsys/components/network/src/WiFi.c:235]: and here

File: wifi-subsys/components/network/src/WiFi.c [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+    int is_configured = wifi_credentials_are_configured();
+
+
```
Comment (luisan00) [inline wifi-subsys/components/network/src/WiFi.c:238]: more

File: wifi-subsys/components/network/src/WiFi.c [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+    ESP_ERROR_CHECK(init_ap());
+    #endif
+
+    #if CONFIG_ENABLE_STA
+    ESP_ERROR_CHECK(esp_wifi_set_mode(WIFI_MODE_STA));
+    ESP_ERROR_CHECK(init_sta());
+    #endif
+
+    return ESP_OK;
+}
```
Comment (luisan00) [inline wifi-subsys/components/network/src/WiFi.c:261]: new line at EOF

File: wifi-subsys/components/network/src/WiFi.h (hunk 3 above)
Comment (luisan00) [inline wifi-subsys/components/network/src/WiFi.h:96]: new line at EOF

File: wifi-subsys/components/network/test/TestWiFi.c [hunk 12]
```diff
@@ -14,9 +14,15 @@
...
-/**
- * @brief       Dummy file , Part of Network component
- *
- * @copyright   Copyright (c) 2022 Mesh for all
- * @author      luisan00 <luisan00@hotmail.com>
- */
+#include "unity.h"
+#include "wifi.h"
+
+
```
Comment (luisan00) [inline wifi-subsys/components/network/test/TestWiFi.c:20]: extra line...

File: wifi-subsys/components/network/src/WiFi.c (hunk 11 above)
Comment (xkevin190) [reply wifi-subsys/components/network/src/WiFi.c:261]: Screen Shot 2022-01-27 at 4 49 38 PM

this was already added

File: wifi-subsys/main/default_params.h [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+#ifndef DEFAULT_PARAMS_H
+#define DEFAULT_PARAMS_H
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#define NVS_APP_NAMESPACE "m4a-firmware"           /*!< NVS application namespace */
+#define NVS_IS_CONFIGURED_KEY "is_configured"       /*!< Is wifi configured? */
```
Comment (luisan00) [inline wifi-subsys/main/default_params.h:33]: group by functionality please

Comment (luisan00) [reply wifi-subsys/main/default_params.h:33]: the rest of file too

File: wifi-subsys/main/default_params.h [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+#define STA_WIFI_SSID "m4a-wifi"           /*!< default STA ssid */
+#define STA_WIFI_PASSWORD "m4a-default"    /*!< default STA password */
+
+#define ESP_MAXIMUM_RETRY 5
+
+#define AP_WIFI_SSID "m4a_AP"           /*!< default AP ssid */
+#define AP_WIFI_PASSWORD "m4a-default"    /*!< default AP password */
+
+
```
Comment (luisan00) [inline wifi-subsys/main/default_params.h:57]: two more

File: wifi-subsys/main/default_params.h (hunk 5 above)
Comment (luisan00) [inline wifi-subsys/main/default_params.h:62]: new line at EOF

**Review by luisan00 — CHANGES_REQUESTED** [review]
round two!

File: wifi-subsys/components/network/src/WiFi.c (hunk 11 above)
Comment (luisan00) [reply wifi-subsys/components/network/src/WiFi.c:261]: Captura de pantalla de 2022-01-27 21-58-14

but not syncronized between your pc and the repo for some reason

**Comment by xkevin190** [comment]
all the requested changes were uploaded, please please approve it :)

File: wifi-subsys/components/network/src/WiFi.c (hunk 9 above)
Comment (luisan00) [inline wifi-subsys/components/network/src/WiFi.c]: extra line

File: wifi-subsys/components/network/test/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,3 @@
+idf_component_register(SRC_DIRS "."
+                       INCLUDE_DIRS "."
+                       REQUIRES network storage unity)
```
Comment (luisan00) [inline wifi-subsys/components/network/test/CMakeLists.txt:3]: new line at EOF

File: wifi-subsys/main/Kconfig.projbuild [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+        default 4
+        help
+            Max number of the STA connects to AP.
+
+    config ENABLE_AP
+        bool "enable AP"
+        default n
+        help
+            enable AP mode
+endmenu
```
Comment (luisan00) [inline wifi-subsys/main/Kconfig.projbuild:57]: add new line here too

File: wifi-subsys/components/network/test/CMakeLists.txt (hunk 15 above)
Comment (xkevin190) [reply wifi-subsys/components/network/test/CMakeLists.txt:3]: Screen Shot 2022-01-27 at 5 19 47 PM

File: wifi-subsys/components/network/src/WiFi.c (hunk 9 above)
Comment (xkevin190) [reply wifi-subsys/components/network/src/WiFi.c]: Screen Shot 2022-01-27 at 5 20 24 PM

where?

**Review by luisan00 — CHANGES_REQUESTED** [review]
round 3

**Comment by xkevin190** [comment]
the new line at EOF lines not showing by github.

**Comment by luisan00** [comment]
Looks good man, please do a squash and go ahead !

**Review by luisan00 — APPROVED** [review]

**Pull request merged**